src/omv/ports/stm32/jpeg.c

```diff
@@ -0,0 +1,753 @@
+
+        if (avail < space) {
+            fb_alloc_fail();
+        }
+
+        dst->size = IMLIB_IMAGE_MAX_SIZE(avail - space);
+        dst->data = fb_alloc(dst->size, FB_ALLOC_PREFER_SIZE | FB_ALLOC_CACHE_ALIGN);
+    }
+
+    if (src->is_compressed) {
+        return true;
+    }
+
+    // Compute size of the APP0 header with cache alignment padding.
+    int app0_size = sizeof(JPEG_APP0);
+    int app0_unalign_size = app0_size % __SCB_DCACHE_LINE_SIZE;
+    int app0_padding_size = app0_unalign_size ? (__SCB_DCACHE_LINE_SIZE - app0_unalign_size) : 0;
+    int app0_total_size = app0_size + app0_padding_size;
+
+    if (dst->size < app0_total_size) {
+        return true; // overflow
+    }
+
+    // Adjust JPEG size and address by app0 header size.
+    dst->size -= app0_total_size;
+    uint8_t *dma_buffer = dst->data + app0_total_size;
+
+    // Destination is too small.
+    if (dst->size < (OUTPUT_CHUNK_SIZE * 2)) {
+        return true; // overflow
+    }
+
+    JPEG_out_data_length_max = dst->size;
+    JPEG_out_data_length = 0;
+    JPEG_input_paused = false;
+    JPEG_output_paused = false;
+
+    uint8_t *mcu_row_buffer = fb_alloc(src_w_mcus_bytes_2, FB_ALLOC_PREFER_SPEED | FB_ALLOC_CACHE_ALIGN);
+
+    for (int y_offset = 0; y_offset < src->h; y_offset += MCU_H) {
+        uint8_t *mcu_row_buffer_ptr = mcu_row_buffer + (src_w_mcus_bytes * ((y_offset / MCU_H) % 2));
+
+        int dy = src->h - y_offset;
+        if (dy > MCU_H) {
+            dy = MCU_H;
+        }
+
+        for (int x_offset = 0; x_offset < src->w; x_offset += MCU_W) {
+            int8_t *Y0 = (int8_t *) (mcu_row_buffer_ptr + (mcu_size * (x_offset / MCU_W)));
+            int8_t *CB = Y0 + JPEG_444_GS_MCU_SIZE;
+            int8_t *CR = CB + JPEG_444_GS_MCU_SIZE;
+
+            int dx = src->w - x_offset;
+            if (dx > MCU_W) {
+                dx = MCU_W;
+            }
+
+            // Copy 8x8 MCUs.
+            jpeg_get_mcu(src, x_offset, y_offset, dx, dy, Y0, CB, CR);
+        }
+
+        // Flush the MCU row for DMA...
+        SCB_CleanDCache_by_Addr((uint32_t *) mcu_row_buffer_ptr, src_w_mcus_bytes);
+
+        if (!y_offset) {
+            // Invalidate the output buffer.
+            SCB_InvalidateDCache_by_Addr(dma_buffer, OUTPUT_CHUNK_SIZE);
+            // Start the DMA process off on the first row of MCUs.
+            HAL_JPEG_Encode_DMA(&JPEG_Handle, mcu_row_buffer_ptr, src_w_mcus_bytes, dma_buffer, OUTPUT_CHUNK_SIZE);
+        } else {
+
+            // Wait for the last row MCUs to be processed before starting the next row.
+            while (!JPEG_input_paused) {
+                __WFI();
+
+                if (JPEG_output_paused) {
+                    memset(&JPEG_Config, 0, sizeof(JPEG_ConfTypeDef));
+                    HAL_JPEG_Abort(&JPEG_Handle);
+                    fb_free(); // mcu_row_buffer (after DMA is aborted)
+                    return true; // overflow
+                }
+            }
```

Comment:
We should use timeout and poll events for all of those loops.
```suggestion
            for (mp_uint_t tickstart = mp_hal_ticks_ms(); !JPEG_input_paused; ) {
                if (JPEG_output_paused || (mp_hal_ticks_ms() - tickstart) > DECODE_TIMEOUT) {
                    jpeg_overflow = true;
                    goto exit_cleanup;
                }
                MICROPY_EVENT_POLL_HOOK;
            }
```

Reply:
Done, also, cleaned up loop exits using goto.